### PR TITLE
fix(Core/Spells): Resolve invalid spell casts on dead players when the cast is valid.

### PR DESCRIPTION
**Pull request by SubstituteR**
## Changes Proposed:

This PR proposes changes to:
-  [x] Core (units, players, creatures, game systems).
-  [ ] Scripts (bosses, spell scripts, creature scripts).
-  [ ] Database (SAI, creatures, etc).

There were two bugs present.
1. Ghosts (ex: Deathly Usher) were unable to detect other ghosts, fixed [here](https://github.com/SubstituteR/azerothcore-wotlk/blob/issues-20509/src/server/game/Entities/Creature/Creature.cpp#L1206).
2. Spells that did not require a dead player (which should allow casts on dead players) were not properly being checked, fixed [here](https://github.com/SubstituteR/azerothcore-wotlk/blob/21e7bcb46534a144eb960501d88787ad4382ffde/src/server/game/Spells/SpellInfo.cpp#L1778-L1782).

These changes were tested against Deathly Usher.

## Issues Addressed:

- 
Closes https://github.com/azerothcore/azerothcore-wotlk/issues/20509
Closes https://github.com/chromiecraft/chromiecraft/issues/2401
Closes https://github.com/azerothcore/azerothcore-wotlk/issues/9356

## SOURCE:

The changes have been validated through:
- [ ] Live research (checked on live servers, e.g Classic WotLK, Retail, etc.)
- [ ] Sniffs (remember to share them with the open source community!)
- [ ] Video evidence, knowledge databases or other public sources (e.g forums, Wowhead, etc.)
- [ ] The changes promoted by this pull request come partially or entirely from another project (cherry-pick). **Cherry-picks must be committed using the proper --author tag in order to be accepted, thus crediting the original authors, unless otherwise unable to be found**

## Tests Performed:

This PR has been:
- [x] Tested in-game by the author.
- [ ] Tested in-game by other community members/someone else other than the author/has been live on production servers.
- [x] This pull request requires further testing and may have edge cases to be tested.

## How to Test the Changes:

- [x] This pull request can be tested by following the reproduction steps provided in the linked issue
- [x] This pull request requires further testing. Provide steps to test your changes. If it requires any specific setup e.g multiple players please specify it as well.

1. Need to verify that enabling two-way visibility between ghosts ([line](https://github.com/SubstituteR/azerothcore-wotlk/blob/issues-20509/src/server/game/Entities/Creature/Creature.cpp#L1206)) does not create problems with other npcs / scripts / etc.

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by TheSCREWEDSoftware** [comment]
Tested all the possiblities reflected on this PR via the issues and description. @Nyeriah can you review the cpp?

File: src/server/game/Entities/Creature/Creature.cpp [hunk 1]
```diff
@@ -1201,8 +1201,10 @@ bool Creature::Create(ObjectGuid::LowType guidlow, Map* map, uint32 phaseMask, u
         m_serverSideVisibilityDetect.SetValue(SERVERSIDE_VISIBILITY_GHOST, GHOST_VISIBILITY_GHOST);
     }
     else if (cinfo->type_flags & CREATURE_TYPE_FLAG_VISIBLE_TO_GHOSTS) // Xinef: Add ghost visibility for ghost units
+    {
         m_serverSideVisibility.SetValue(SERVERSIDE_VISIBILITY_GHOST, GHOST_VISIBILITY_ALIVE | GHOST_VISIBILITY_GHOST);
-
+        m_serverSideVisibilityDetect.SetValue(SERVERSIDE_VISIBILITY_GHOST, GHOST_VISIBILITY_ALIVE | GHOST_VISIBILITY_GHOST);
```
Comment (heyitsbench) [inline src/server/game/Entities/Creature/Creature.cpp:1206]: Please remove the brackets surrounding this line.
https://www.azerothcore.org/wiki/cpp-code-standards#brackets

File: src/server/game/Spells/SpellInfo.cpp [hunk 2]
```diff
@@ -1775,13 +1775,11 @@ SpellCastResult SpellInfo::CheckTarget(Unit const* caster, WorldObject const* ta
             return SPELL_FAILED_TARGET_AFFECTING_COMBAT;
 
         // only spells with SPELL_ATTR3_ONLY_ON_GHOSTS can target ghosts
-        if (((IsRequiringDeadTarget() != 0) != unitTarget->HasAuraType(SPELL_AURA_GHOST)) && !(IsDeathPersistent() && IsAllowingDeadTarget()))
-        {
-            if (AttributesEx3 & SPELL_ATTR3_ONLY_ON_GHOSTS)
```
Comment (sudlud) [inline src/server/game/Spells/SpellInfo.cpp:1780]: What about spells with this attribute - do they still work correctly?

Comment (SubstituteR) [reply src/server/game/Spells/SpellInfo.cpp:1780]: Responded on Discord

Comment (Kitzunu) [reply src/server/game/Spells/SpellInfo.cpp:1780]: IsAllowingDeadTarget() checks that attribute. It was a double check before

File: src/server/game/Spells/SpellInfo.cpp [hunk 3]
```diff
@@ -1775,13 +1775,11 @@ SpellCastResult SpellInfo::CheckTarget(Unit const* caster, WorldObject const* ta
...
-                return SPELL_FAILED_TARGET_NOT_GHOST;
-            else
-                return SPELL_FAILED_BAD_TARGETS;
-        }
+        if (IsRequiringDeadTarget() && !unitTarget->HasAuraType(SPELL_AURA_GHOST))
+            return SPELL_FAILED_TARGET_NOT_GHOST;
+
+        if (IsDeathPersistent() && IsAllowingDeadTarget())
```
Comment (Kitzunu) [inline src/server/game/Spells/SpellInfo.cpp:1781]: These checks should be reversed
```suggestion
        if (!IsDeathPersistent() && !IsAllowingDeadTarget())
```

Comment (SubstituteR) [reply src/server/game/Spells/SpellInfo.cpp]: I read it as blocking the spell if the death was persistent but the spell was allowed on a dead target.  

I'll use your suggestion.

**Review by Kitzunu — APPROVED** [review]
lgtm

**Comment by Kitzunu** [comment]
thanks for the pr

**Pull request merged**